scripts/main.py

```diff
@@ -37,15 +41,74 @@ def run(config: dict, cache: str):
     incident_model = LinearIncidentUptakeModel().fit(
         incident_train_data, grouping_factors
     )
-    cumulative_projections = incident_model.predict(
-        config["timeframe"]["start"],
-        config["timeframe"]["end"],
-        config["timeframe"]["interval"],
-        grouping_factors,
-    )
-    print(cumulative_projections)
-    incident_projections = cumulative_projections.to_incident(grouping_factors)
-    print(incident_projections)
+
+    option = config["option"]
+
+    if option == "projection":
+        cumulative_projections = incident_model.predict(
+            config["timeframe"]["start"],
+            config["timeframe"]["end"],
+            config["timeframe"]["interval"],
+            grouping_factors,
+        )
+        print(cumulative_projections)
+        incident_projections = cumulative_projections.to_incident(grouping_factors)
+        print(incident_projections)
+
+    elif option == "evaluation":
+        # Make projections sequentially
+        interval_str = config["timeframe"]["interval"]
+        interval_number = int(interval_str[0])
+
+        dates = pl.date_range(
+            config["timeframe"]["start"],
+            config["timeframe"]["end"] - dt.timedelta(interval_number),
+            config["timeframe"]["interval"],
+            eager=True,
+        )
+
+        scores = pl.DataFrame(
+            schema={
+                "forecast_start": pl.Date,
+                "forecast_end": pl.Date,
+                "score": pl.Float64,
+                "type": pl.String,
+            }
+        )
+
+        for date in dates:
+            # Generate cumulative predictions
+            cumulative_projections = incident_model.predict(
+                date,
+                config["timeframe"]["end"],
+                config["timeframe"]["interval"],
+                grouping_factors,
+            )
+
+            pred = iup.PointForecast(
+                cumulative_projections.to_incident(grouping_factors)
+            )
```

Comment:
This is a good point. Here I evaluated retrospective forecasting, which is evaluating the model performance as the forecast period changes. This requires the model stays the same to objectively observe if the model forecasts better when changing the forecast time length, lead time (forecast start). Thus the training data needs to be the same to have the same model. While adding data to training sequentially as forecasting moving forward is also valid, which is called real-time forecasting(I think?). This is more commonly applied in operational forecasting in practice. Which method to use depends on the scope of this project, I'd like to hear about what @swo thinks!

Reply:
To make sure we're on the same page, some definitions:

- Forecast date: The date the forecast was run. In a retrospective analysis, you run a forecast with the data that would have been available on this date.
- Observation date: The date that the thing you're trying to forecast occurred.
- Forecast horizon: Time between the forecast date and the observation date. E.g., if a Jan 1 forecast is predicting what will happen on Feb 1, that's a 1 month horizon.

For a given forecast date, estimates on observation dates that are further into the future (e.g., at longer forecast horizons) will have lower accuracy.

So I think @eschrom is saying that, in a retrospective analysis, you run forecasts on many forecast dates. Later forecast dates have access to more data. The data that comes out of a retrospective analysis will be many combinations of forecast dates and observation dates. Then you would score a model by saying, e.g., what was its accuracy at a 1-month forecast horizon, averaged over all the relevant forecast dates.

![image](https://github.com/user-attachments/assets/5c07bd18-f0b6-4581-93ac-56a7eb271e33)

Does this help at all?

Reply:
I had not been familiar with the vocabulary, so this helps a lot. I like the Paint-style sketch, too ;)

"In a retrospective analysis, you run a forecast with the data that would have been available on this date ...  in a retrospective analysis, you run forecasts on many forecast dates. Later forecast dates have access to more data" This is exactly what I am saying, but more eloquent.

Reply:
But how do we objectively evaluate the performance if continuously adding data to training? In this way, the forecasts made on later forecast dates must be more accurate than the ones made on early forecast dates only because they have more training data, and we cannot evaluate the relation between forecast performance and forecast dates (such as if approaching a seasonal onset affect performance) unbiasedly.

Reply:
You are correct that "forecasts made on later forecast dates must be more accurate than the ones made on early forecast dates only because they have more training data." But this is not a problem. 

We want to compare forecasts between _models_, not between _forecast dates_. We don't care that a given model always performs better for later forecast dates. We care whether one model performs better than another at a given forecast date.

Here's a sketch of my own:
![image](https://github.com/user-attachments/assets/d3ef3ae1-399a-43c8-bbbe-13f71e52fca4)

Both models improve in performance as forecast date advances. But the key is to compare the models. In this case, we see that Model 2 performs better for early forecast dates, but Model 1 performs better for later forecast dates.

Reply:
See [b127d43](https://github.com/CDCgov/cfa-immunization-uptake-projection/pull/82/commits/b127d43fe50e80b1f31a684d7a3c6e0596265c17)

Reply:
This adjustment looks right! I think this is the crucial thing to fix in this PR.